@vates/nbd-client/index.mjs

```diff
@@ -0,0 +1,273 @@
+    await this.#writeInt32(NBD_FLAG_FIXED_NEWSTYLE) // client also support  NBD_FLAG_C_FIXED_NEWSTYLE
+
+    if (this.#useSecureConnection) {
+      // upgrade socket to TLS if needed
+      await this.#sendOption(NBD_OPT_STARTTLS)
+      await this.#tlsConnect()
+    }
+
+    // send export name we want to access.
+    // it's  implictly closing the negociation phase.
+    await this.#write(Buffer.from(OPTS_MAGIC))
+    await this.#writeInt32(NBD_OPT_EXPORT_NAME)
+    await this.#writeInt32(this.#exportname.length)
+    await this.#write(Buffer.from(this.#exportname))
+
+    // 8 (export size ) + 2 (flags) + 124 zero = 134
+    // must read all to ensure nothing stays in the  buffer
+    const answer = await this.#read(134)
+    this.#exportSize = answer.readBigUInt64BE(0)
+    const transmissionFlags = answer.readInt16BE(8)
+    assert(transmissionFlags & NBD_FLAG_HAS_FLAGS, 'NBD_FLAG_HAS_FLAGS') // must always be 1 by the norm
+
+    // note : xapi server always send NBD_FLAG_READ_ONLY (3) as a flag
+  }
+
+  #takeFromBuffer(length) {
+    const res = Buffer.from(this.#receptionBuffer.slice(0, length))
+    this.#receptionBuffer = this.#receptionBuffer.slice(length)
+    return res
+  }
+
+  #read(length) {
+    // if we got enough data in local cache : return them immediatly
+    if (this.#receptionBuffer.length >= length) {
+      return this.#takeFromBuffer(length)
+    }
+    // if not, wait for the data, add the read to the fifo
+    return new Promise(resolve => {
+      this.#rawReadBacklog.push({
+        resolve,
+        length,
+      })
+    })
+  }
+
+  #write(buffer) {
+    return new Promise(resolve => {
+      this.#serverSocket.write(buffer, resolve)
+    })
+  }
+
+  async #readInt32() {
+    const buffer = await this.#read(4)
+    return buffer.readInt32BE(0)
+  }
+
+  async #readInt64() {
+    const buffer = await this.#read(8)
+    return buffer.readBigUInt64BE(0)
+  }
+
+  #writeInt32(int) {
+    const buffer = Buffer.alloc(4)
+    buffer.writeInt32BE(int)
+    return this.#write(buffer)
+  }
+
+  async #handleBlockReadResponse() {
+    const magic = await this.#readInt32()
+
+    if (magic !== NBD_REPLY_MAGIC) {
+      throw new Error(`magic number for block answer is wrong : ${magic}`)
+    }
+
+    const error = await this.#readInt32()
+    if (error !== 0) {
+      // @todo use error code from constants.mjs
+      throw new Error(`GOT ERROR CODE  : ${error}`)
+    }
+
+    const blockQueryId = await this.#readInt64()
+    const query = this.#commandQueryBacklog[blockQueryId]
+    if (!query) {
+      throw new Error(` no query associated with id ${blockQueryId} ${Object.keys(this.#commandQueryBacklog)}`)
+    }
+    delete this.#commandQueryBacklog[blockQueryId]
+    const data = await this.#read(query.size)
+    query.resolve(data)
+  }
+
+  async readBlock(index, size = NBD_DEFAULT_BLOCK_SIZE) {
+    const queryId = this.#nextCommandQueryId
+    this.#nextCommandQueryId++
```

Comment:
This is what postfix increment is for :slightly_smiling_face: 

```suggestion
    const queryId = this.#nextCommandQueryId++
```

Reply:
I think it does not improve readability